index.js

```diff
@@ -270,35 +287,78 @@ class ExecutorQueue extends Executor {
      * @async  _start
      * @param  {Object} config               Configuration
      * @param  {Object} [config.annotations] Optional key/value object
      * @param  {String} config.build         Build object
      * @param  {Array}  config.blockedBy     Array of job IDs that this job is blocked by. Always blockedby itself
+     * @param  {Array}  config.freezeWindows Array of cron expressions that this job cannot run during
      * @param  {String} config.apiUri        Screwdriver's API
      * @param  {String} config.jobId         JobID that this build belongs to
      * @param  {String} config.buildId       Unique ID for a build
+     * @param  {String} config.eventId       eventID that this build belongs to
+     * @param  {Object} config.pipeline      Pipeline of the job
+     * @param  {Fn}     config.tokenGen      Function to generate JWT from username, scope and scmContext
      * @param  {String} config.container     Container for the build to run in
      * @param  {String} config.token         JWT to act on behalf of the build
      * @return {Promise}
      */
     async _start(config) {
         await this.connect();
-        const { build, buildId, jobId, blockedBy } = config;
+        const { build, buildId, jobId, blockedBy, freezeWindows } = config;
+
+        if (!this.tokenGen) {
+            this.tokenGen = config.tokenGen;
+        }
 
         delete config.build;
-        // Store the config in redis
-        await this.redisBreaker.runCommand('hset', this.buildConfigTable,
-            buildId, JSON.stringify(config));
 
-        // Note: arguments to enqueue are [queue name, job name, array of args]
-        const enq = await this.queueBreaker.runCommand('enqueue', this.buildQueue, 'start', [{
-            buildId,
-            jobId,
-            blockedBy: blockedBy.toString()
-        }]);
+        const currentTime = new Date();
+        const origTime = new Date(currentTime.getTime());
+
+        timeOutOfWindows(freezeWindows, currentTime);
+
+        let enq;
+
+        if (currentTime.getTime() > origTime.getTime()) {
+            await this.queueBreaker.runCommand('delDelayed', this.frozenBuildQueue,
+                'startFrozen', [{
+                    jobId
+                }]);
+
+            await this.redisBreaker.runCommand('hset', this.frozenBuildTable,
+                jobId, JSON.stringify(config));
+
+            enq = await this.queueBreaker.runCommand('enqueueAt', currentTime.getTime(),
+                this.frozenBuildQueue, 'startFrozen', [{
+                    jobId
+                }]
+            );
+        } else {
+            // Store the config in redis
+            await this.redisBreaker.runCommand('hset', this.buildConfigTable,
+                buildId, JSON.stringify(config));
+
+            // Note: arguments to enqueue are [queue name, job name, array of args]
+            enq = await this.queueBreaker.runCommand('enqueue', this.buildQueue, 'start', [{
```

Comment:
This case apply to the case where there is no freeze window OR it was previously frozen and now not anymore?
For the 2nd case, don't we also need to remove it from `frozenBuildTable` and `frozenBuildQueue`?

Reply:
If there is no freeze window or if the current time doesn't fall into a freeze window it will run the else condition. Agreed about the 2nd case, will implement it

Reply:
Another question related to this, want to clarify the flow
executor-queue.start(build with freeze window) -> put in freeze queue -> wait until it's unfreezed -> new event start. But the original build status is still `QUEUED` since it's never put into the build queue. We should probably update the status to `FREEZED`. Otherwise it can be confusing

Reply:
Fixed